ui/src/app/utils/cluster-initializer.spec.tsx

```diff
@@ -0,0 +1,305 @@
+let mockStartCluster: SpyInstance;
+
+const baseCluster: Cluster = {
+  clusterName: 'aou-rw-3',
+  clusterNamespace: 'aou-rw-12345',
+  status: ClusterStatus.Running
+};
+
+describe('ClusterInitializer', () => {
+
+  beforeEach(() => {
+    jest.useFakeTimers();
+
+    registerApiClient(ClusterApi, new ClusterApiStub());
+    registerApiClientNotebooks(NotebooksClusterApi, new NotebooksClusterApiStub());
+
+    mockGetCluster = jest.spyOn(clusterApi(), 'getCluster');
+    mockCreateCluster = jest.spyOn(clusterApi(), 'createCluster');
+    mockDeleteCluster = jest.spyOn(clusterApi(), 'deleteCluster');
+    mockStartCluster = jest.spyOn(notebooksClusterApi(), 'startCluster');
+  });
+
+  afterEach(() => {
+    jest.useRealTimers();
+    jest.clearAllMocks();
+  });
+
+  const mockGetClusterCalls = (responses: Array<Cluster>) => {
+    for (const response of responses) {
+      mockGetCluster.mockImplementationOnce((workspaceNamespace) => {
+        return response;
+      });
+    }
+  };
+
+  const createInitializer = (options?: Partial<ClusterInitializerOptions>): ClusterInitializer => {
+    return new ClusterInitializer({
+      workspaceNamespace: 'aou-rw-12345',
+      ...options,
+    });
+  };
+
+  const runTimersUntilSettled = async(p: Promise<any>) => {
+    let isSettled = false;
+    p.then(() => isSettled = true).catch((e) => {
+      isSettled = true;
+      // Re-throw the error.
+      // throw e;
+    });
+    while (!isSettled) {
+      await new Promise(setImmediate);
+      jest.runAllTimers();
+    }
+  };
+
+  const runInitializerAndTimers = async(initializer: ClusterInitializer): Promise<Cluster> => {
+    const clusterPromise = initializer.run();
+    await runTimersUntilSettled(clusterPromise);
+    return await clusterPromise;
+  };
+
+  it('should resolve promise if cluster is in ready state', async() => {
+    // This tests the simplest case of the initializer. No polling necessary.
+    mockGetClusterCalls([baseCluster]);
+    const initializer = createInitializer();
+    const cluster = await initializer.run();
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('should resume cluster if it is initially stopped', async() => {
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Stopped},
+      // Here is where we expect, chronologically, a call to startCluster. The next two
+      // cluster statuses mock out what we might expect to see after that call.
+      {...baseCluster, status: ClusterStatus.Starting},
+      {...baseCluster, status: ClusterStatus.Running}
+    ]);
+    const initializer = createInitializer();
+    const cluster = await runInitializerAndTimers(initializer);
+    expect(mockStartCluster).toHaveBeenCalled();
+    expect(cluster.status).toEqual(ClusterStatus.Running);
+  });
+
+  it('should call callback with cluster status', async() => {
+    mockGetClusterCalls([
+      {...baseCluster, status: ClusterStatus.Stopped},
```

Comment:
nit: these `mockGetClusterCalls()` invocations might be readable with a helper lambda/method like `baseWith(s: ClusterStatus): Cluster`. So
`{...baseCluster, status: ClusterStatus.Starting}` becomes `baseWith(ClusterStatus.Starting)`.

It's just a lot of `...baseCluster`, but I'm sure some of that distraction is culture shock.